### ISSUE TITLE
How to check docker daemon options when using daemon.json

### ISSUE THREAD

**Issue by praving5** [docker/docker#21559]
**Output of `docker version`:**

```
# docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:39:25 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:39:25 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 6
Server Version: 1.10.3
Storage Driver: devicemapper
 Pool Name: docker-253:1-67129347-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 2.381 GB
 Data Space Total: 107.4 GB
 Data Space Available: 40.16 GB
 Metadata Space Used: 3.523 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.144 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-12-01)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.18.17-13.el7.x86_64
Operating System: Red Hat Enterprise Linux
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 7.812 GiB
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Steps to reproduce the issue:**
1. Create `/etc/docker/daemon.json` file with one option, say `"icc": false`
2. Start the docker daemon - `systemctl start docker.service`
3. Execute `ps -ef | grep docker`

You will not see --icc=false parameter passed to docker daemon but actually docker daemon is running with that parameter. How can we list all the configuration parameters that docker daemon is currently running with? I think there should be a uniform way to do that instead of relying on ps -ef. A simple command such as `docker daemon info` might help to give out a json file with all the options that the daemon is currently running with. This double way of configuring docker daemon options creates a severe auditing problem - there is no standard and official way to find out docker daemon configuration parameters. The parameters passed via daemon.json are 'hidden' from `ps -ef`.

**Describe the results you received:**
`ps -ef` only reports parameters that are passed via `/usr/lib/systemd/system/docker.service`. There is no standard command that reports all the docker daemon options that are currently loaded.

**Describe the results you expected:**
I expected that `ps -ef | grep docker` would list all the docker daemon configuration options irrespective of whether the options are passed in `/usr/lib/systemd/system/docker.service` or `/etc/docker/daemon.json`.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by praving5** [comment]
Anyone on this yet?

**Comment by thaJeztah** [comment]
I brought this up recently in another PR; https://github.com/docker/docker/pull/20410#issuecomment-196768891, but we haven't had time yet to give it much thought (busy with the 1.11 release)

**Comment by calavera** [comment]
> How to check docker daemon options when using daemon.json 

`cat /etc/docker/daemon.json` :sparkles:

But seriously, these kind of things are why configuration management tools exist.

**Comment by emmetog** [comment]
Although not a show stopper, this would be handy. I'm tweaking the dns docker daemon options and because it's not working as I expect it to it would be nice to be able to see which options the daemon has currently loaded.

**Comment by twang2218** [comment]
Yes, it would be great to have a command which can dump the current effective daemon config, or dump the config at beginning of the daemon logs when the daemon started. That would helped the diagnosis a lot.

**Comment by ryanlyy** [comment]
yes. that will be great for debugging too.

**Comment by LukaszStem** [comment]
Is there any update on this?

**Comment by galindro** [comment]
any return guys?

**Comment by cpuguy83** [comment]
No update. I think it would be great to be able to pull the merged configuration from the API.

**Comment by twang2218** [comment]
I created a PR #30363 to dump the configuration in the log after its loaded, on both engine starting or configuration reloading. I think it will help the diagnosis.

**Comment by kyprifog** [comment]
I also think some work could be done around this.  I updated /etc/docker/daemon.json expecting it to update the docker daemon config but could only do so through the UI and could not easily figure out where the UI config was being saved so I could update that instead.

**Comment by praving55** [comment]
@thaJeztah - This has been pending over a year now. Could this be prioritized in some way? It is still hard to believe that we don't have a single command to display the loaded daemon configuration.

**Comment by HYmian** [comment]
The Same issue when I use docker 1.12.6

**Comment by thaJeztah** [comment]
@praving55 it's still on our radar (and actually brought this up last week); current thoughts were to have an API endpoint that exposes the active configuration, but input/proposals are welcome. There's also this issue that may have to be taken into account for the design; https://github.com/moby/moby/issues/30892

**Comment by anuragpaliwal80** [comment]
Is this is still pending or do we have the solution to dump all flags details now ?

**Comment by praving55** [comment]
Nothing that I have seen in 2 years.

**Comment by thaJeztah** [comment]
Small improvement; when live-reloading the configuration, this PR will print the active configuration in the logs; https://github.com/moby/moby/pull/36019

**Comment by Katiyman** [comment]
This req is very handy kindly let me know if this is possible in docker 18.x

**Comment by tvdinh** [comment]
A command to dump all "effective" current configuration would be great.

**Comment by jirislav** [comment]
No progress yet? :crying_cat_face:

**Comment by JoeAlamo** [comment]
Anything on this? very hard to determine whether changes you make to config have been implemented or not..

**Comment by akunaatrium** [comment]
How's it progressing?

**Comment by BenasPaulikas** [comment]
issue still open but it's already working. 
`docker info`

**Comment by thaJeztah** [comment]
`docker info` always existed, but doesn't show all options that are configurable, so the ticket is still open because of that

**Comment by BenasPaulikas** [comment]
Ah ok. I've changed live-reload in config file and that was visible in `docker info`. ignore my previous comment then.

**Comment by mouund** [comment]
Hello, 
unfortunately still unable to see the actual configuration sources, it can be really annoying to have a clear idea of which file/command line args are used. More than that it doesn't show what the daemon.json brings to the equation in ps;

> $ ps -ef | grep docker
> root        8350       1  0 19:58 ?        00:00:00 /usr/bin/dockerd --containerd=/run/containerd/containerd.sock
> root        8936    5620  0 20:03 pts/0    00:00:00 grep --color=auto docker

lsof doesn't give you the config file which is loaded

> $ lsof -p  8350 | grep daemon


> $ cat /etc/docker/daemon.json 
> {
>   "hosts": [ "unix:///var/run/docker.sock"]
> }

Can be a good QOL improvement to either have the args list in the ps output or be able to see which file is currently loaded as conf

**Comment by don-rumata** [comment]
```bash
dockerd --validate
```
?
